libyul/backends/evm/Dominator.h

```diff
@@ -0,0 +1,276 @@
+		// Finds ancestor with lowest semi-dominator dfs number (i.e. index).
+		auto eval = [&](size_t _v) -> size_t
+		{
+			if (ancestor[_v] != std::numeric_limits<size_t>::max())
+			{
+				compressPath(ancestor, label, semi, _v);
+				return label[_v];
+			}
+			return _v;
+		};
+
+		auto toIdx = [&](Vertex const& v) { return m_vertexIndex[v]; };
+
+		// step 1
+		std::set<Vertex> visited;
+		// predecessors(w): The set of vertices ``v`` such that (``v``, ``w``) is an edge of the graph.
+		std::vector<std::set<size_t>> predecessors(numVertices);
+		// bucket(w): a set of vertices whose semidominator is ``w``
+		// The index of the array represents the vertex's ``dfIdx``
+		std::vector<std::deque<size_t>> bucket(numVertices);
+		// idom(w): the index of the immediate dominator of ``w``
+		std::vector<size_t> idom(numVertices, std::numeric_limits<size_t>::max());
+		// The number of vertices reached during the dfs.
+		// The vertices are indexed based on this number.
+		size_t dfIdx = 0;
+		auto dfs = [&](Vertex const& _v, auto _dfs) -> void {
+			if (visited.count(_v))
+				return;
+			visited.insert(_v);
+			m_vertex[dfIdx] = _v;
+			m_vertexIndex[_v] = dfIdx;
+			semi[dfIdx] = dfIdx;
+			label[dfIdx] = dfIdx;
+			dfIdx++;
+			ForEachSuccessor{}(_v, [&](Vertex const& w) {
+				if (semi[dfIdx] == std::numeric_limits<size_t>::max())
+				{
+					parent[dfIdx] = m_vertexIndex[_v];
+					_dfs(w, _dfs);
+				}
+				predecessors[m_vertexIndex[w]].insert(m_vertexIndex[_v]);
+			});
+		};
+		dfs(_entry, dfs);
+
+		// Process the vertices in decreasing order of the dfs number
+		for (size_t w: m_vertex | ranges::views::reverse | ranges::views::transform(toIdx))
+		{
+			// step 3
+			// NOTE: this is an optimization, i.e. performing the step 3 before step 2.
+			// The goal is to process the bucket in the beginning of the loop for the vertex ``w``
+			// instead of ``parent[w]`` in the end of the loop as described in the original paper.
+			// Inverting those steps ensures that a bucket is only processed once and
+			// it does not need to be erased.
+			// The optimization proposal is available here: https://jgaa.info/accepted/2006/GeorgiadisTarjanWerneck2006.10.1.pdf pg.77
+			ranges::for_each(
+				bucket[w],
+				[&](size_t v)
+				{
+					size_t u = eval(v);
+					idom[v] = (semi[u] < semi[v]) ? u : w;
+				}
+			);
+
+			// step 2
+			for (size_t v: predecessors[w])
+			{
+				size_t u = eval(v);
+				if (semi[u] < semi[w])
+					semi[w] = semi[u];
+			}
+			bucket[semi[w]].emplace_back(w);
+			link(parent[w], w);
+		}
+
+		// step 4
+		idom[0] = 0;
+		for (size_t w: m_vertex | ranges::views::drop(1) | ranges::views::transform(toIdx))
+			if (idom[w] != semi[w])
+				idom[w] = idom[idom[w]];
+
+		return idom;
+	}
+private:
+	// Keep the list of vertices in the dfs order.
+	// i.e. m_vertex[i]: the vertex whose dfs index is i.
+	std::vector<Vertex> m_vertex;
+	// Maps Vertex to their dfs index.
+	std::map<Vertex, size_t> m_vertexIndex;
```

Comment:
The problem is that without a map from `Vertex` to index we will need to perform a search in the `m_vertices` vector every time we need to use/know the index of a specific vertex. For example, this is needed when querying if a vertex is dominated or dominates any other vertex and when computing the predecessors.

Reply:
Ah, ok. I guess initially I misunderstood what it was for, because I did not see that `Vertex` was meant to be an ID and not some kind of heavier payload (using the latter for lookups just wasn't making sense). Now I see that this is just a bi-directional map between two layers of IDs. This really circles back to the issue I pointed out in https://github.com/ethereum/solidity/pull/14460#discussion_r1291531794.

In that case it's fine, though maybe switching to using `Vertex` as a proper ID would on the other hand make `m_vertex` obsolete?

Reply:
No, it is the opposite indeed. `Vertex`, now renamed to `V`, was never meant to be an ID, but the vertex object itself. But I need to have a way to lookup what ID a vertex object is associated with. And the ID is only assigned to the vertex during the DFS. So the idea was that `V` would implement `operator<` allowing them to be used as key in the index lookup map. In case of complex objects I imagine that this would be done with some sort of internal ID or hashing. In the current tests I just used a lexicographic order based on the Vertex label (i.e. `TestVertex.name`).

Maybe we could use `boost::MultiIndex` to allow lookup by ID and by Vertex. But I guess we can refactor that later. The main idea there is to be able to access a vertex by their ID and to be able to retrieve their ID based on the object itself.

Reply:
No, it is the opposite indeed. `Vertex`, now renamed to `V`, was never meant to be an ID, but the vertex object itself. I need to have a way to lookup what ID a vertex object is associated with. And the ID is only assigned to the vertex during the DFS (i.e. the ID, index of the vector, represents the DFS number assigned to the vertex when the graph is traversed). So the idea was that `V` would implement `operator<` allowing them to be used as key in the index lookup map. In case of complex objects I imagine that this would be done with some sort of internal ID or hashing. In the current tests I just used a lexicographic order based on the Vertex label (i.e. `TestVertex.name`).

Maybe we could use `boost::MultiIndex` to allow lookup by ID and by Vertex. But I guess we can refactor that later. The main idea there is to be able to access a vertex by their ID and to be able to retrieve their ID based on the object itself.

But yeah, you are right and I could indeed make the index of the vector `m_vertices` to not be the same as the DFS index (which is what I use as ID), which I believe is what you are suggesting, and I would keep the map from vector indexes to DFS numbers.